lib/Language/languageHelpers.ts

```diff
@@ -1,15 +1,25 @@
-import i18next from "i18next";
+import { i18n } from "i18next";
+
+export const TRANSLATE_KEY_PREFIX = "translate#";
 
 /**
- * Takes a given string and translates it if it exists, otherwise return
+ * Takes a given string and translates it if it's a translation key that exists,
+ * otherwise returns the string.
+ *
+ * This function doesn't use the global i18next instance because this function is predominately
+ * used in the UIMode and using React components should use withTranslation HOC or useTranslation
+ * hook to ensure that they update when the aop language changes.
+ *
+ * @param keyOrString Either a prefixed translation key or a string
+ * @param i18n An i18next instance
  */
-export function useTranslationIfExists(keyOrString: string) {
-  if (keyOrString && keyOrString.indexOf("translate#") === 0) {
-    const translationKey = keyOrString.substr("translate#".length);
-    return i18next.exists(translationKey)
-      ? i18next.t(translationKey)
+export function useTranslationIfExists(keyOrString: string, i18n: i18n) {
+  if (keyOrString.indexOf(TRANSLATE_KEY_PREFIX) === 0) {
```

Comment:
If I recall correctly it can happen that `keyOrString` be undefined (specified by user)

Reply:
It shouldn't, since the type signature doesn't include `undefined`, but the function was being called wrongly by a .jsx file (`lib/ReactViews/Map/Panels/HelpPanel/HelpPanelItem.jsx `) with a parameter that was sometimes undefined. That's also been fixed in this PR

Reply:
I just run a check on `trainerItems` if the title is undefined it will throw an error although it is typescript file
```json5
"trainerItems": [
  {
    //"title": "stepbystepguide.gettingstarted.title", // this will throw rror
    "footnote": "stepbystepguide.gettingstarted.footnote",
    "steps": [
      {
        //"title": "stepbystepguide.gettingstarted.step1.title", // this will throw error when opening steps guide
        "markdownDescription": "stepbystepguide.gettingstarted.step1.markdownDescription"
      }
    ]
  }
]
```

Reply:
@steve9164 which way do you want to go on this?

Reply:
Probably there's an implicit `any` in the code that deals with `trainerItems` - anything that deals with JSON. You could fix it by explicitly checking existence of each property of the JSON object. Using `unknown` helps out here - e.g. `fetchJson(...) as unknown` and then `if (o.title) {...}`

Reply:
Perhaps this is better:

```ts
export function applyTranslationIfExists(
  keyOrString: string,
  i18n: i18n,
  options?: { [key: string]: string }
): string {
  // keyOrString could be undefined in some cases even if we type it as string
  if (isJsonString(keyOrString as unknown)) {
    if (keyOrString.indexOf(TRANSLATE_KEY_PREFIX) === 0) {
      const translationKey = keyOrString.substr(TRANSLATE_KEY_PREFIX.length);
      return i18n.exists(translationKey)
        ? i18n.t(translationKey, options)
        : translationKey;
    } else {
      return keyOrString;
    }
  }
  // Return an empty string if keyOrString was undefined. All functions that call this require a string to be returned
  return "";
}
```